File: js/task3c/S3_PIPE_05.js
const pipe = (...fns) => value =>
  fns.reduce((v, fn) => fn(v), value);

const parse = line => {
  const [level, rest] = line.split(": ");
  return { level, user: rest.split("=")[1] };
};

const getInfoUsers = pipe(
  lines => lines.map(parse),
  items => items.filter(i => i.level === "INFO"),
  items => items.map(i => i.user)
);

console.log(
  getInfoUsers([
    "INFO: user=42",
    "ERROR: user=99",
    "INFO: user=7"
  ])
);

File: py/task3c/S3_PIPE_05.py
def parse_line(line):
    try:
        level, rest = line.split(":", 1)
        parts = dict(p.split("=") for p in rest.strip().split())
        return {"level": level, **parts}
    except:
        return None


def extract_users(lines):
    result = []
    for line in lines:
        parsed = parse_line(line)
        if parsed and parsed.get("level") == "INFO":
            try:
                result.append(int(parsed["user"]))
            except:
                pass
    return result


logs = [
    "INFO: user=42 action=login",
    "ERROR: user=x",
    "INFO: user=7 action=logout"
]

print(extract_users(logs))
